src/timely-util/src/panic.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::panic;
+
+use mz_ore::halt;
+
+/// Intercepts [`timely_communication`] panics and downgrades them to [`halt`]s.
+pub fn halt_on_timely_communication_panic() {
+    let old_hook = panic::take_hook();
+    panic::set_hook(Box::new(move |panic_info| {
+        match panic_info.payload().downcast_ref::<String>() {
+            Some(e) if e.starts_with("timely communication error:") => {
+                halt!("{}", e);
```

Comment:
this is...subtle and we should add a comment here explaining what is going on: we are not catching a panic at all, we are "beating the unwind" to the punch by `std::process::exit`-ing beforehand

Reply:
Yes! What is subtle about it?

Reply:
I feel like that's already captured by the docstring on this function, though! ("Intercepts expected timely communication panics and downgrades them to halts.") I'd like to avoid duplicating an explanation of what `halt!` does here, as there's a big documentation comment on the `halt!` macro already, which feels like the right place!

I did beef up the function's documentation comment a bit to mention the rationale for this change—and I called out why we have to do this via string matching instead of downcasting.